pygfx/helpers/_gizmo.py

```diff
@@ -0,0 +1,397 @@
+        )
+        rotate_xy = gfx.Mesh(
+            sphere_geo,
+            gfx.MeshBasicMaterial(color="#0000ff"),
+        )
+        on_arc = halfway * 2**0.5 / 2
+        rotate_yz.position.set(0, on_arc, on_arc)
+        rotate_zx.position.set(on_arc, 0, on_arc)
+        rotate_xy.position.set(on_arc, on_arc, 0)
+
+        # ---
+
+        # Rotate objects to their correct orientation
+        for ob in [line_y, translate_y, scale_y, arc_zx]:
+            ob.rotation.set_from_axis_angle(Vector3(0, 0, 1), np.pi / 2)
+        for ob in [line_z, translate_z, scale_z, arc_xy]:
+            ob.rotation.set_from_axis_angle(Vector3(0, -1, 0), np.pi / 2)
+
+        # Store the objectss
+        self._line_children = line_x, line_y, line_z
+        self._arc_children = arc_yz, arc_zx, arc_xy
+        self._translate_children = (
+            translate_x,
+            translate_y,
+            translate_z,
+            translate_screen,
+        )
+        self._scale_children = scale_x, scale_y, scale_z
+        self._rotate_children = rotate_yz, rotate_zx, rotate_xy
+
+        # Assign dimensions
+        for triplet in [
+            self._line_children,
+            self._translate_children[:3],
+            self._scale_children,
+            self._rotate_children,
+        ]:
+            for i, ob in enumerate(triplet):
+                ob.dim = i
+        translate_screen.dim = "screen"
+
+        # Attach to the gizmo object
+        self.add(*self._line_children)
+        self.add(*self._arc_children)
+        self.add(translate_screen, *self._translate_children)
+        self.add(*self._scale_children)
+        self.add(*self._rotate_children)
+
+    def update_matrix_world(self, *args, **kwargs):
+        # This gets called by the renderer just before rendering.
+        self._adjust_to_camera()
+        super().update_matrix_world(*args, **kwargs)
+
+    def _adjust_to_camera(self):
+        if not (self._object_to_control and self._canvas and self._camera):
+            return
+        camera = self._camera
+        canvas_size = self._canvas.get_logical_size()
+
+        # Get center position (of the gizmo) in world and screen coords
+        center_world = self._object_to_control.position
+        center_screen = center_world.clone().project(camera)
+
+        # Get how our direction vectors express on screen
+        pos1 = center_screen.clone().add(Vector3(100, 0, 0)).unproject(camera)
+        pos2 = center_screen.clone().add(Vector3(0, 100, 0)).unproject(camera)
+        pos1.multiply_scalar(2 / 100 / canvas_size[0])
+        pos2.multiply_scalar(2 / 100 / canvas_size[1])
+        self._screen_vecs = pos1, pos2
+
+        # Determine the scale of this object
+        avg_length = 0.5 * (
+            self._screen_vecs[0].length() + self._screen_vecs[1].length()
+        )
+        scale_scalar = self._screen_size * avg_length
+        scale = [scale_scalar, scale_scalar, scale_scalar]
```

Comment:
It's this. It's a bit hard to explain :)  we transform a horizontal and vertical vector in screen space to scene space. The average of *these* vectors is used to scale the widget, so that it approximately takes up the intended screen space.